emacs/.config/emacs/config.org
#+title: Emacs configuration with Org Mode
#+author: Jahor Laryn
#+email: jahor@tuta.io
#+startup: noinlineimages

* Preface
Poor man's lisp machine config.
* Init
** Proper beginning
#+begin_src emacs-lisp
  ;;; Commentary:
  ;;; Emacs config by Jahor Laryn.
  ;;; This file was automatically generated by `org-babel-tangle'.
  ;;;
  ;;; Code:
#+end_src

** Minimal use-package setup
#+begin_src emacs-lisp
  (setq-default gc-cons-threshold 402653184
                gc-cons-percentage 0.6
                inhibit-compacting-font-caches t
                message-log-max 16384
                file-name-handler-alist nil)

  (require 'package)
  (setq package-enable-at-startup nil)
  (setq package-check-signature nil)
  (add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
  (package-initialize)
  (unless (package-installed-p 'use-package)
    (package-refresh-contents)
    (package-install 'use-package))
  (require 'use-package)
  (setq use-package-always-ensure t)
#+end_src
* Custom functions
#+begin_src emacs-lisp
  (defun zzz ()
    "Suspend to ram"
    (interactive)
    (shell-command "systemctl suspend"))
  (defun web-view ()
    (interactive)
    (shell-command (concat "firefox " buffer-file-name)))
#+end_src
* Keybindings
#+begin_src emacs-lisp
;; Terminal
(global-set-key "\C-x\ a" '(lambda ()(interactive)(eshell)))
(global-set-key "\C-x\ c" '(lambda ()(interactive)(find-file "~/.config/emacs/config.org")))
#+end_src
* Better defaults
** Startup
#+begin_src emacs-lisp
  (setq inhibit-splash-screen t)
#+end_src
** Encoding
#+begin_src emacs-lisp
  (prefer-coding-system 'utf-8)
  (when (display-graphic-p)
    (setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING)))
#+end_src
** Layout
#+begin_src emacs-lisp
  (setq default-input-method 'russian-computer)
#+end_src
** Dired
#+begin_src emacs-lisp
;; iji
#+end_src

** Encoding
#+begin_src emacs-lisp
  (set-language-environment "UTF-8")
#+end_src
** Tabline
#+begin_src emacs-lisp
(when (and window-system
             (not (version< emacs-version "27")))
    (use-package tab-line
      ;;:hook (after-init . global-tab-line-mode)
      :config
      (defun tab-line-close-tab (&optional e)
        "Close the selected tab.

  If tab is presented in another window, close the tab by using
  `bury-buffer` function.  If tab is unique to all existing
  windows, kill the buffer with `kill-buffer` function.  Lastly, if
  no tabs left in the window, it is deleted with `delete-window`
  function."
        (interactive "e")
        (let* ((posnp (event-start e))
               (window (posn-window posnp))
               (buffer (get-pos-property 1 'tab (car (posn-string posnp)))))
          (with-selected-window window
            (let ((tab-list (tab-line-tabs-window-buffers))
                  (buffer-list (flatten-list
                                (seq-reduce (lambda (list window)
                                              (select-window window t)
                                              (cons (tab-line-tabs-window-buffers) list))
                                            (window-list) nil))))
              (select-window window)
              (if (> (seq-count (lambda (b) (eq b buffer)) buffer-list) 1)
                  (progn
                    (if (eq buffer (current-buffer))
                        (bury-buffer)
                      (set-window-prev-buffers window (assq-delete-all buffer (window-prev-buffers)))
                      (set-window-next-buffers window (delq buffer (window-next-buffers))))
                    (unless (cdr tab-list)
                      (ignore-errors (delete-window window))))
                (and (kill-buffer buffer)
                     (unless (cdr tab-list)
                       (ignore-errors (delete-window window)))))))))

      (defcustom tab-line-tab-min-width 10
        "Minimum width of a tab in characters."
        :type 'integer
        :group 'tab-line)

      (defcustom tab-line-tab-max-width 30
        "Maximum width of a tab in characters."
        :type 'integer
        :group 'tab-line)

      (defun aorst/tab-line-name-buffer (buffer &rest _buffers)
        "Create name for tab with padding and truncation.

  If buffer name is shorter than `tab-line-tab-max-width' it gets
  centered with spaces, otherwise it is truncated, to preserve
  equal width for all tabs.  This function also tries to fit as
  many tabs in window as possible, so if there are no room for tabs
  with maximum width, it calculates new width for each tab and
  truncates text if needed.  Minimal width can be set with
  `tab-line-tab-min-width' variable."
        (with-current-buffer buffer
          (let* ((window-width (window-width (get-buffer-window)))
                 (close-button-size (if tab-line-close-button-show
                                        (length (substring-no-properties tab-line-close-button))
                                      0))
                 (tab-amount (length (tab-line-tabs-window-buffers)))
                 (window-max-tab-width (/ window-width tab-amount))
                 (tab-width (- (cond ((>= window-max-tab-width tab-line-tab-max-width)
                                      tab-line-tab-max-width)
                                     ((< window-max-tab-width tab-line-tab-min-width)
                                      tab-line-tab-min-width)
                                     (t window-max-tab-width))
                               close-button-size))
                 (buffer-name (string-trim (buffer-name)))
                 (name-width (length buffer-name)))
            (if (>= name-width (- tab-width 3))
                (concat  " " (truncate-string-to-width buffer-name (- tab-width 3)) "… ")
              (let* ((padding (make-string (/ (- tab-width name-width) 2) ?\s))
                     (buffer-name (concat padding buffer-name))
                     (name-width (length buffer-name)))
                (concat buffer-name (make-string (- tab-width name-width) ?\s)))))))

      (setq tab-line-close-button-show t
            tab-line-new-button-show nil
            tab-line-separator ""
            tab-line-tab-name-function #'aorst/tab-line-name-buffer
            tab-line-right-button (propertize (if (char-displayable-p ?▶) " ▶ " " > ")
                                              'keymap tab-line-right-map
                                              'mouse-face 'tab-line-highlight
                                              'help-echo "Click to scroll right")
            tab-line-left-button (propertize (if (char-displayable-p ?◀) " ◀ " " < ")
                                             'keymap tab-line-left-map
                                             'mouse-face 'tab-line-highlight
                                             'help-echo "Click to scroll left")
            tab-line-close-button (propertize (if (char-displayable-p ?×) "× " "x ")
                                              'keymap tab-line-tab-close-map
                                              'mouse-face 'tab-line-close-highlight
                                              'help-echo "Click to close tab"))

      (let ((bg (if (facep 'solaire-default-face)
                    (face-attribute 'solaire-default-face :background)
                  (face-attribute 'default :background)))
            (fg (face-attribute 'default :foreground))
            (base (face-attribute 'mode-line :background))
            (box-width (/ (line-pixel-height) 2)))
        (when (and (color-defined-p bg)
                   (color-defined-p fg)
                   (color-defined-p base)
                   (numberp box-width))
          
              (set-face-attribute 'tab-line nil
                                  :background base
                                  :foreground fg
                                  :height 1.0
                                  :inherit nil
                                  :box (list :line-width -1 :color base))

              (set-face-attribute 'tab-line-tab nil
                                  :foreground fg
                                  :background bg
                                  :weight 'normal
                                  :inherit nil
                                  :box (list :line-width 5 :color bg))

              (set-face-attribute 'tab-line-tab-inactive nil
                                  :foreground fg
                                  :background base
                                  :weight 'normal
                                  :inherit nil
                                :box (list :line-width 5 :color base))

              (set-face-attribute 'tab-line-tab-current nil
                                  :foreground fg
                                  :background bg
                                  :weight 'normal
                                  :inherit nil
                                :box (list :line-width 5 :color bg))

))
      (setq tab-line-exclude-modes '())
      (dolist (mode '(ediff-mode
                      process-menu-mode
                      term-mode
                      vterm-mode))
        (add-to-list 'tab-line-exclude-modes mode))

      (defun aorst/tab-line-drop-caches ()
        "Drops `tab-line' cache in every window."
        (dolist (window (window-list))
          (set-window-parameter window 'tab-line-cache nil)))

      (add-hook 'window-configuration-change-hook #'aorst/tab-line-drop-caches)))



#+end_src
** Tab character
#+begin_src emacs-lisp
  (setq-default indent-tabs-mode nil)
#+end_src

** Font
#+begin_src emacs-lisp
  (when (member "Iosevka" (font-family-list))
    (set-face-attribute 'default nil :font "Iosevka 25"))
#+end_src
** Yes or No
#+begin_src emacs-lisp
    (fset 'yes-or-no-p 'y-or-n-p)
#+end_src
** Org-mode
#+begin_src emacs-lisp
  (setq org-src-tab-acts-natively t)
#+end_src
** Useless shit
*** Menus
#+begin_src emacs-lisp
(scroll-bar-mode 0)
(tool-bar-mode 0)
(menu-bar-mode 0)
(blink-cursor-mode 0)
(setq use-dialog-box nil)
#+end_src
*** Disabled commands
#+begin_src emacs-lisp
  (put 'set-goal-column 'disabled nil)
  (put 'narrow-to-region 'disabled nil)
  (put 'upcase-region 'disabled nil)
  (put 'downcase-region 'disabled nil)
  (put 'scroll-left 'disabled nil)
  (put 'erase-buffer 'disabled nil)
#+end_src
*** Backups
#+begin_src emacs-lisp
(setq backup-by-copying t)
(setq create-lockfiles nil)
(setq backup-directory-alist '(("." . "/tmp")))
(setq auto-save-file-name-transforms '((".*" "/tmp/" t)))
#+end_src
** Default Major Mode
I would  like =*scratch*= buffer  to use  =fundamental= mode, so  it loads
faster.  Since  it's no  longer Emacs  Lisp interaction  buffer, let's
also remove initial message:

#+begin_src emacs-lisp
    (setq initial-major-mode 'fundamental-mode)
    (setq initial-scratch-message "")
#+end_src


** Line numbers
#+begin_src emacs-lisp
;;  (add-hook 'prog-mode-hook (lambda () (display-line-numbers-mode 1)))
  (use-package display-line-numbers
    :custom
    (display-line-numbers-grow-only t)
    (display-line-numbers-width-start t))
#+end_src
** Paren
#+begin_src emacs-lisp
   (setq show-paren-delay 0)
#+end_src

** Prog Mode
This isn't really a configuration of  Prog Mode itself, but some hooks
that change how  programming related modes behave. One  feature that I
think is really important, especially  when working with lisp code, is
ability  to see  matching bracket  when cursor  stands near  the other
bracket. So I enable it for every programming language.

#+begin_src emacs-lisp
  (add-hook 'prog-mode-hook (lambda ()
                              (show-paren-mode 1)
                              (display-line-numbers-mode 1)
                              (hl-line-mode 1)))
#+end_src
** Frame
#+begin_src emacs-lisp
  (when window-system
    (setq window-divider-default-right-width 7)
    (window-divider-mode 1)
    (set-face-attribute 'window-divider nil
                        :foreground (face-attribute
                                     'mode-line-inactive :background)))
#+end_src
** Title
Emacs uses weird method of naming a window, I'd like to see a file I'm
currently working  on in the task  bar in case my  Emacs was minimized
for some reason.

#+begin_src emacs-lisp
  (setq-default frame-title-format '("%b — Emacs"))
#+end_src
** Mouse and Scrolling
I don't want menus on shift clicks:

#+begin_src emacs-lisp
  (global-unset-key (kbd "S-<down-mouse-1>"))
  (global-unset-key (kbd "<mouse-3>"))
  (global-set-key [mouse-3] menu-bar-edit-menu)
  (global-unset-key (kbd "S-<mouse-3>"))
#+end_src
** Selection
#+begin_src emacs-lisp
  (delete-selection-mode t)
#+end_src
** History
#+begin_src emacs-lisp
  (savehist-mode 1)
#+end_src
** Bell
I'm not  trying to  insult on anyone,  but usage of  the bell  is just
insane. It's  not 1980's, computers  have their own speakers  and rich
displays, why would anyone want  to use builtin speaker?  Why everyone
should know when I'm mistaken? Disable bell.

#+begin_src emacs-lisp
  (setq ring-bell-function 'ignore)
#+end_src


* Third-party
** Which-key
#+begin_src emacs-lisp
  (use-package which-key
    :config
    (which-key-mode t)
    (setq which-key-show-early-on-C-h t)
    (setq which-key-idle-delay 0)
    (setq which-key-idle-secondary-delay 0)
    :ensure t)
#+end_src
** Theme
#+begin_src emacs-lisp
     ;; Tomorrow
     ;; (use-package color-theme-sanityinc-tomorrow
     ;; :config
     ;; (load-theme 'sanityinc-tomorrow-blue t))

    ;; Spacemacs

    ;; (use-package spacemacs-theme
    ;;   :defer t
    ;;   :init (load-theme 'spacemacs-dark t))
   (use-package doom-themes
    :config
    (load-theme 'doom-one t)
    ;; Enable flashing mode-line on errors
    (doom-themes-visual-bell-config)
    (setq doom-themes-enable-bold t)
    (setq doom-themes-treemacs-enable-variable-pitch nil)
    ;; Corrects (and improves) org-mode's native fontification.
    (doom-themes-org-config)

    ;; Enable custom treemacs theme (all-the-icons must be installed!)
    (doom-themes-treemacs-config))

#+end_src

** Hydra
#+begin_src emacs-lisp
  (use-package hydra)
#+end_src
** Ivy
Ivy  is a  narrowing  framework  like Helm,  but  much  lighter in  my
experience.  It  integrates with  Counsel that handles  minibuffer, so
let's install it too.

#+begin_src emacs-lisp
  (use-package ivy
    :commands ivy-mode
    :bind (("C-x b" . ivy-switch-buffer)
           ("C-x C-b" . ivy-switch-buffer))
    :custom
    (ivy-mode 1))

  (use-package counsel
    :commands (counsel-M-x
               counsel-find-file
               counsel-file-jump
               counsel-recentf
               counsel-rg
               counsel-describe-function
               counsel-describe-variable
               counsel-find-library)
    :bind (("M-x" . counsel-M-x)
            ("C-x C-f" . counsel-find-file)
           ("C-x f" . counsel-file-jump)
           ("C-x C-r" . counsel-recentf)
           ("C-x d" . counsel-dired)
           ("C-h f" . counsel-describe-function)
           ("C-h C-f" . counsel-describe-face)
           ("C-h v" . counsel-describe-variable)
           ("C-h l" . counsel-find-library)))

  (use-package swiper
    :bind (("C-S-s" . swiper)
           ("C-S-r" . swiper)))
#+end_src

** Company
Complete anything framework. Nothing much to say. Does it's job.

#+begin_src emacs-lisp
  (use-package company
    :bind (:map company-active-map
           ("TAB" . company-complete-common-or-cycle)
           ("<tab>" . company-complete-common-or-cycle)
           ("<S-Tab>" . company-select-previous)
           ("<backtab>" . company-select-previous)
           ("C-n" . company-select-next)
           ("C-p" . company-select-previous))
    :hook (after-init . global-company-mode)
    :custom
    (company-require-match 'never)
    (company-minimum-prefix-length 0)
    (company-tooltip-align-annotations t)

emacs/.config/emacs/config.el
;;; Commentary:
;;; Emacs config by Jahor Laryn.
;;; This file was automatically generated by `org-babel-tangle'.
;;;
;;; Code:

(setq-default gc-cons-threshold 402653184
              gc-cons-percentage 0.6
              inhibit-compacting-font-caches t
              message-log-max 16384
              file-name-handler-alist nil)

(require 'package)
(setq package-enable-at-startup nil)
(setq package-check-signature nil)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
(package-initialize)
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

(defun zzz ()
  "Suspend to ram"
  (interactive)
  (shell-command "systemctl suspend"))
(defun web-view ()
  (interactive)
  (shell-command (concat "firefox " buffer-file-name)))

;; Terminal
(global-set-key "\C-x\ a" '(lambda ()(interactive)(eshell)))
(global-set-key "\C-x\ c" '(lambda ()(interactive)(find-file "~/.config/emacs/config.org")))

(setq inhibit-splash-screen t)

(prefer-coding-system 'utf-8)
(when (display-graphic-p)
  (setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING)))

(setq default-input-method 'russian-computer)

;; iji

(set-language-environment "UTF-8")

(when (and window-system
             (not (version< emacs-version "27")))
    (use-package tab-line
      ;;:hook (after-init . global-tab-line-mode)
      :config
      (defun tab-line-close-tab (&optional e)
        "Close the selected tab.

  If tab is presented in another window, close the tab by using
  `bury-buffer` function.  If tab is unique to all existing
  windows, kill the buffer with `kill-buffer` function.  Lastly, if
  no tabs left in the window, it is deleted with `delete-window`
  function."
        (interactive "e")
        (let* ((posnp (event-start e))
               (window (posn-window posnp))
               (buffer (get-pos-property 1 'tab (car (posn-string posnp)))))
          (with-selected-window window
            (let ((tab-list (tab-line-tabs-window-buffers))
                  (buffer-list (flatten-list
                                (seq-reduce (lambda (list window)
                                              (select-window window t)
                                              (cons (tab-line-tabs-window-buffers) list))
                                            (window-list) nil))))
              (select-window window)
              (if (> (seq-count (lambda (b) (eq b buffer)) buffer-list) 1)
                  (progn
                    (if (eq buffer (current-buffer))
                        (bury-buffer)
                      (set-window-prev-buffers window (assq-delete-all buffer (window-prev-buffers)))
                      (set-window-next-buffers window (delq buffer (window-next-buffers))))
                    (unless (cdr tab-list)
                      (ignore-errors (delete-window window))))
                (and (kill-buffer buffer)
                     (unless (cdr tab-list)
                       (ignore-errors (delete-window window)))))))))

      (defcustom tab-line-tab-min-width 10
        "Minimum width of a tab in characters."
        :type 'integer
        :group 'tab-line)

      (defcustom tab-line-tab-max-width 30
        "Maximum width of a tab in characters."
        :type 'integer
        :group 'tab-line)

      (defun aorst/tab-line-name-buffer (buffer &rest _buffers)
        "Create name for tab with padding and truncation.

  If buffer name is shorter than `tab-line-tab-max-width' it gets
  centered with spaces, otherwise it is truncated, to preserve
  equal width for all tabs.  This function also tries to fit as
  many tabs in window as possible, so if there are no room for tabs
  with maximum width, it calculates new width for each tab and
  truncates text if needed.  Minimal width can be set with
  `tab-line-tab-min-width' variable."
        (with-current-buffer buffer
          (let* ((window-width (window-width (get-buffer-window)))
                 (close-button-size (if tab-line-close-button-show
                                        (length (substring-no-properties tab-line-close-button))
                                      0))
                 (tab-amount (length (tab-line-tabs-window-buffers)))
                 (window-max-tab-width (/ window-width tab-amount))
                 (tab-width (- (cond ((>= window-max-tab-width tab-line-tab-max-width)
                                      tab-line-tab-max-width)
                                     ((< window-max-tab-width tab-line-tab-min-width)
                                      tab-line-tab-min-width)
                                     (t window-max-tab-width))
                               close-button-size))
                 (buffer-name (string-trim (buffer-name)))
                 (name-width (length buffer-name)))
            (if (>= name-width (- tab-width 3))
                (concat  " " (truncate-string-to-width buffer-name (- tab-width 3)) "… ")
              (let* ((padding (make-string (/ (- tab-width name-width) 2) ?\s))
                     (buffer-name (concat padding buffer-name))
                     (name-width (length buffer-name)))
                (concat buffer-name (make-string (- tab-width name-width) ?\s)))))))

      (setq tab-line-close-button-show t
            tab-line-new-button-show nil
            tab-line-separator ""
            tab-line-tab-name-function #'aorst/tab-line-name-buffer
            tab-line-right-button (propertize (if (char-displayable-p ?▶) " ▶ " " > ")
                                              'keymap tab-line-right-map
                                              'mouse-face 'tab-line-highlight
                                              'help-echo "Click to scroll right")
            tab-line-left-button (propertize (if (char-displayable-p ?◀) " ◀ " " < ")
                                             'keymap tab-line-left-map
                                             'mouse-face 'tab-line-highlight
                                             'help-echo "Click to scroll left")
            tab-line-close-button (propertize (if (char-displayable-p ?×) "× " "x ")
                                              'keymap tab-line-tab-close-map
                                              'mouse-face 'tab-line-close-highlight
                                              'help-echo "Click to close tab"))

      (let ((bg (if (facep 'solaire-default-face)
                    (face-attribute 'solaire-default-face :background)
                  (face-attribute 'default :background)))
            (fg (face-attribute 'default :foreground))
            (base (face-attribute 'mode-line :background))
            (box-width (/ (line-pixel-height) 2)))
        (when (and (color-defined-p bg)
                   (color-defined-p fg)
                   (color-defined-p base)
                   (numberp box-width))
          
              (set-face-attribute 'tab-line nil
                                  :background base
                                  :foreground fg
                                  :height 1.0
                                  :inherit nil
                                  :box (list :line-width -1 :color base))

              (set-face-attribute 'tab-line-tab nil
                                  :foreground fg
                                  :background bg
                                  :weight 'normal
                                  :inherit nil
                                  :box (list :line-width 5 :color bg))

              (set-face-attribute 'tab-line-tab-inactive nil
                                  :foreground fg
                                  :background base
                                  :weight 'normal
                                  :inherit nil
                                :box (list :line-width 5 :color base))

              (set-face-attribute 'tab-line-tab-current nil
                                  :foreground fg
                                  :background bg
                                  :weight 'normal
                                  :inherit nil
                                :box (list :line-width 5 :color bg))

))
      (setq tab-line-exclude-modes '())
      (dolist (mode '(ediff-mode
                      process-menu-mode
                      term-mode
                      vterm-mode))
        (add-to-list 'tab-line-exclude-modes mode))

      (defun aorst/tab-line-drop-caches ()
        "Drops `tab-line' cache in every window."
        (dolist (window (window-list))
          (set-window-parameter window 'tab-line-cache nil)))

      (add-hook 'window-configuration-change-hook #'aorst/tab-line-drop-caches)))

(setq-default indent-tabs-mode nil)

(when (member "Iosevka" (font-family-list))
  (set-face-attribute 'default nil :font "Iosevka 25"))

(fset 'yes-or-no-p 'y-or-n-p)

(setq org-src-tab-acts-natively t)

(scroll-bar-mode 0)
(tool-bar-mode 0)
(menu-bar-mode 0)
(blink-cursor-mode 0)
(setq use-dialog-box nil)

(put 'set-goal-column 'disabled nil)
(put 'narrow-to-region 'disabled nil)
(put 'upcase-region 'disabled nil)
(put 'downcase-region 'disabled nil)
(put 'scroll-left 'disabled nil)
(put 'erase-buffer 'disabled nil)

(setq backup-by-copying t)
(setq create-lockfiles nil)
(setq backup-directory-alist '(("." . "/tmp")))
(setq auto-save-file-name-transforms '((".*" "/tmp/" t)))

(setq initial-major-mode 'fundamental-mode)
(setq initial-scratch-message "")

;;  (add-hook 'prog-mode-hook (lambda () (display-line-numbers-mode 1)))
  (use-package display-line-numbers
    :custom
    (display-line-numbers-grow-only t)
    (display-line-numbers-width-start t))

(setq show-paren-delay 0)

(add-hook 'prog-mode-hook (lambda ()
                            (show-paren-mode 1)
                            (display-line-numbers-mode 1)
                            (hl-line-mode 1)))

(when window-system
  (setq window-divider-default-right-width 7)
  (window-divider-mode 1)
  (set-face-attribute 'window-divider nil
                      :foreground (face-attribute
                                   'mode-line-inactive :background)))

(setq-default frame-title-format '("%b — Emacs"))

(global-unset-key (kbd "S-<down-mouse-1>"))
(global-unset-key (kbd "<mouse-3>"))
(global-set-key [mouse-3] menu-bar-edit-menu)
(global-unset-key (kbd "S-<mouse-3>"))

(delete-selection-mode t)

(savehist-mode 1)

(setq ring-bell-function 'ignore)

(use-package which-key
  :config
  (which-key-mode t)
  (setq which-key-show-early-on-C-h t)
  (setq which-key-idle-delay 0)
  (setq which-key-idle-secondary-delay 0)
  :ensure t)

;; Tomorrow
  ;; (use-package color-theme-sanityinc-tomorrow
  ;; :config
  ;; (load-theme 'sanityinc-tomorrow-blue t))

 ;; Spacemacs

 ;; (use-package spacemacs-theme
 ;;   :defer t
 ;;   :init (load-theme 'spacemacs-dark t))
(use-package doom-themes
 :config
 (load-theme 'doom-one t)
 ;; Enable flashing mode-line on errors
 (doom-themes-visual-bell-config)
 (setq doom-themes-enable-bold t)
 (setq doom-themes-treemacs-enable-variable-pitch nil)
 ;; Corrects (and improves) org-mode's native fontification.
 (doom-themes-org-config)

 ;; Enable custom treemacs theme (all-the-icons must be installed!)
 (doom-themes-treemacs-config))

(use-package hydra)

(use-package ivy
  :commands ivy-mode
  :bind (("C-x b" . ivy-switch-buffer)
         ("C-x C-b" . ivy-switch-buffer))
  :custom
  (ivy-mode 1))

(use-package counsel
  :commands (counsel-M-x
             counsel-find-file
             counsel-file-jump
             counsel-recentf
             counsel-rg
             counsel-describe-function
             counsel-describe-variable
             counsel-find-library)
  :bind (("M-x" . counsel-M-x)
          ("C-x C-f" . counsel-find-file)
         ("C-x f" . counsel-file-jump)
         ("C-x C-r" . counsel-recentf)
         ("C-x d" . counsel-dired)
         ("C-h f" . counsel-describe-function)
         ("C-h C-f" . counsel-describe-face)
         ("C-h v" . counsel-describe-variable)
         ("C-h l" . counsel-find-library)))

(use-package swiper
  :bind (("C-S-s" . swiper)
         ("C-S-r" . swiper)))

(use-package company
  :bind (:map company-active-map
         ("TAB" . company-complete-common-or-cycle)
         ("<tab>" . company-complete-common-or-cycle)
         ("<S-Tab>" . company-select-previous)
         ("<backtab>" . company-select-previous)
         ("C-n" . company-select-next)
         ("C-p" . company-select-previous))
  :hook (after-init . global-company-mode)
  :custom
  (company-require-match 'never)
  (company-minimum-prefix-length 0)
  (company-tooltip-align-annotations t)
  (company-frontends '(company-pseudo-tooltip-unless-just-one-frontend
                       ;; company-preview-frontend
                       company-echo-metadata-frontend))
  (company-backends '(company-capf company-files))
  (company-tooltip-minimum-width 30)
  (company-tooltip-maximum-width 60))

(use-package company-posframe
  :after company
  :custom
  (company-posframe-quickhelp-show-header nil)
  (company-posframe-show-indicator nil)
  (company-posframe-show-metadata nil)
  (company-posframe-quickhelp-show-params
   (list :poshandler #'company-posframe-quickhelp-right-poshandler
         :internal-border-width 1
         :timeout 60
         :internal-border-color (face-attribute 'mode-line-inactive :background)
         :no-properties nil
         :poshandler nil))
  :config
  (company-posframe-mode))

(use-package magit
  :bind (("C-x C-g" . magit))
  :config
  (add-hook 'magit-mode-hook (lambda ()
                               (tab-line-mode 0))))

(use-package all-the-icons
  :config
  (when (not (member "all-the-icons" (font-family-list)))
    (all-the-icons-install-fonts t)))
(use-package all-the-icons-dired
  :after all-the-icons dired
  :ensure t
  :config
  (add-hook 'dired-mode-hook 'all-the-icons-dired-mode))

(use-package treemacs
  :commands (treemacs-follow-mode
             treemacs-filewatch-mode
             treemacs-fringe-indicator-mode
             treemacs-load-theme)
  :bind (("<f7>" . treemacs)
         ("<f8>" . treemacs-select-window)))

(use-package tex-site
  :ensure auctex
  :mode ("\\.tex\\'" . latex-mode)
  :config
  (setq TeX-auto-save t)
  (setq TeX-parse-self t)
  (setq-default TeX-master nil)
  )

(use-package haskell-mode
    :config
    ;; (defun hlint ()
    ;;   (interactive)
    ;;   (shell-command "hlint ."))
    ;; (local-set-key (kbd "M-q") 'hlint)
    (setq haskell-process-type 'ghci)
    (add-hook 'haskell-mode-hook (lambda () (interactive-haskell-mode)))
    )
(use-package company-ghci
  :config (push 'company-ghci company-backends))

(use-package nix-mode)

(use-package web-mode
  :config 
  (add-to-list 'auto-mode-alist '("\\.phtml\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.tpl\\.php\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.[agj]sp\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.as[cp]x\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.erb\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.mustache\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.djhtml\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.html?\\'" . web-mode))
  (add-hook 'web-mode-hook (lambda () 
                             (local-set-key (kbd "C-x C-v") 'web-view)
  )))

(use-package multiple-cursors
  :ensure t
  :config
  (setq mc/always-run-for-all t)
  (global-set-key (kbd "C->") 'mc/mark-next-like-this)
  (global-set-key (kbd "C-<") 'mc/mark-previous-like-this)
  (global-unset-key (kbd "C-<down-mouse-1>"))
  (global-set-key (kbd "C-<mouse-1>") 'mc/add-cursor-on-click)
  (global-set-key (kbd "C-c C-<") 'mc/mark-all-like-this))

(use-package avy
  :ensure t
  :bind
  (("C-:" . avy-goto-char)
   ("C-'" . avy-goto-char-2)))

(use-package org-bullets
  :config
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  (setq org-bullets-bullet-list '("◉" "○" "⊕" "⊝" "⍟" "⊜" "⊗")))

(use-package flycheck)

(use-package lsp-mode :commands lsp :ensure t
  :custom
  (ccls-executable "~/std/ccls/Release/ccls")

      (ccls-initialization-options
          '(:clang (:extraArgs ["-isystem/Library/Developer/CommandLineTools/usr/include/c++/v1"
                                "-isystem/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include"
                                "-isystem/usr/local/include"
                                "-isystem/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0/include"
                                "-isystem/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include"
                                "-isystem/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
                                "-isystem/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks"]
                    :resourceDir "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0")))
      :hook
      ((c++-mode c-mode) . lsp)
  )



  ;; (use-package lsp-ui :commands lsp-ui-mode :ensure t)
  ;; (use-package company-lsp
  ;;   :ensure t
  ;;   :commands company-lsp
  ;;   :config (push 'company-lsp company-backends)) 


;; set prefix for lsp-command-keymap (few alternatives - "C-l", "C-c l")
;; (setq lsp-keymap-prefix "s-l")

;; (use-package lsp-mode
;;     :hook (;; replace XXX-mode with concrete major-mode(e. g. python-mode)
;;             (c++-mode . lsp)
;;             ;; if you want which-key integration
;;             (lsp-mode . lsp-enable-which-key-integration))
;;     :custom 
;;     (lsp-clients-clangd-executable "clangd")

;;     :commands lsp)

;; ;; optionally
;; (use-package lsp-ui :commands lsp-ui-mode)
;; ;; if you are helm user
;; (use-package helm-lsp :commands helm-lsp-workspace-symbol)
;; ;; if you are ivy user
;; (use-package lsp-ivy :commands lsp-ivy-workspace-symbol)
;; (use-package lsp-treemacs :commands lsp-treemacs-errors-list)


  ;; (use-package ccls
  ;;   :ensure t
  ;;   :config
  ;;   (setq ccls-executable "ccls")
  ;;   (setq lsp-prefer-flymake nil)
  ;;   (setq-default flycheck-disabled-checkers '(c/c++-clang c/c++-cppcheck c/c++-gcc))


  ;;   (setq ccls-initialization-options
  ;;         '(:clang (:extraArgs ["-isystem/Library/Developer/CommandLineTools/usr/include/c++/v1"
  ;;                               "-isystem/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include"
  ;;                               "-isystem/usr/local/include"
  ;;                               "-isystem/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0/include"
  ;;                               "-isystem/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include"
  ;;                               "-isystem/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
  ;;                               "-isystem/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks"]
  ;;                   :resourceDir "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0")))


  ;;   :hook ((c-mode c++-mode objc-mode) .
  ;;          (lambda () (require 'ccls) (lsp))))

      ;; if you are ivy user
      ;; (use-package lsp-ivy :commands lsp-ivy-workspace-symbol)
      ;; (use-package lsp-treemacs :commands lsp-treemacs-errors-list)


      ;; optional if you want which-key integration
      (use-package which-key
        :config
        (which-key-mode))

(use-package minions
  :commands minions-mode
  :init (minions-mode 1))

(use-package dashboard
  :ensure t
  :config
  (dashboard-setup-startup-hook))
